### ISSUE TITLE
Local MOUSE_WHEEL_CLICK_ANGLE rule not applied

### ISSUE THREAD

**Issue by bdaase** [systemd/systemd#10459]
**systemd version the issue has been seen with**
> systemd 239
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid

**Used distribution**
 > Ubuntu 18.10

**Expected behaviour you didn't see**
 > MOUSE_WHEEL_CLICK_ANGLE rules set in `/etc/udev/hwdb.d/71-mouse-local.hwdb` are applied for a Logitech MX Anywhere 2 and the mouse wheel scroll speed changes.

**Unexpected behaviour you saw**
 > It was not applied

**Steps to reproduce the problem**

0. Have a Logitech MX Anywhere 2
1. Create the `/etc/udev/hwdb.d/71-mouse-local.hwdb` file and fill it with (I experimented with different values for the angle, it didn't make a difference)
```
mouse:usb:v046dpc52b:name:*:
 MOUSE_WHEEL_CLICK_ANGLE=30
```
The `vid/pid` are correct due to `lsusb` which has the following output
```
Bus 004 Device 004: ID 152d:0539 JMicron Technology Corp. / JMicron USA Technology Corp. JMS539/567 SuperSpeed SATA II/III 3.0G/6.0G Bridge
Bus 004 Device 003: ID 0bda:8153 Realtek Semiconductor Corp. RTL8153 Gigabit Ethernet Adapter
Bus 004 Device 002: ID 0424:5807 Standard Microsystems Corp. 
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 006: ID 24ae:2002  
Bus 003 Device 005: ID 046d:c52b Logitech, Inc. Unifying Receiver
Bus 003 Device 004: ID 0bda:4014 Realtek Semiconductor Corp. 
Bus 003 Device 002: ID 0424:2807 Standard Microsystems Corp. 
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 003: ID 0489:e0a2 Foxconn / Hon Hai 
Bus 001 Device 002: ID 0bda:58f4 Realtek Semiconductor Corp. 
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```
3. Run `systemd-hwdb update` and `udevadm trigger /dev/input/event*` as root.
4. Use the scroll wheel of the mouse, nothing changed.

Edit: After making the changes mentioned above I ran `libinput debug-events` and get the following event when scrolling:
`event16  POINTER_AXIS     +14.66s	vert 15.00* horiz 0.00 (wheel)`
which looks for me (the documentation says 15 is the default value) like all changes, even those shipped in `/lib/udev/hwdb.d/70-mouse.hwdb`, are ignored.

**Comment by BjoernDaase** [comment]
This problem was already discussed in this thread https://askubuntu.com/questions/285689/increase-mouse-wheel-scroll-speed (comments of first answer) but they did not find any solution as well.

**Comment by gdamjan** [comment]
from what I can see, ubuntu 18.10 uses libinput 1.12 which no longer uses the hwdb files:
http://who-t.blogspot.com/2018/09/whats-new-in-libinput-112.html

**Comment by keszybz** [comment]
@whot can you comment on the general direction here?

**Comment by BjoernDaase** [comment]
@gdamjan Thank you for this information! I'm trying to have a look on the quirk files later today and convert the hwdb file.

Anyway, could we please add a hint to the hwdb files (or to representative one) that these entries are ignored? I think I'm by far not the only one who didn't find that out....

**Comment by whot** [comment]
The wheel click angle is considered a hw property and still parsed by libinput, nothing has changed for that udev property. The only ones that moved from a hwdb entry to the libinput quirks were libinput-internal properties.

Does `udevadm info /sys/class/input/eventX` for your event node list the property? If not, your match rule doesn't apply and the issue is somewhere in the hwdb. Otherwise, it's in libinput, `sudo libinput debug-events --verbose` may have some log to hint at the issue.

A click angle of 30 seems strange though, that's twice as large a movement as your average mouse. Are you sure you only get 12 clicks for a full rotation?

**Comment by BjoernDaase** [comment]
@whot The 30 is absolutely random. I have also tested with 10, 20 and 60 (to make sure that it's not a problem with a specific value) but it didn't make any difference.
That's my output of `udevadm info /sys/class/input/event12`
```
P: /devices/pci0000:00/0000:00:1c.4/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb3/3-2/3-2.6/3-2.6:1.2/0003:046D:C52B.0016/0003:046D:4063.0017/input/input95/event12
N: input/event12
S: input/by-id/usb-Logitech_USB_Receiver-if02-event-mouse
S: input/by-path/pci-0000:39:00.0-usb-0:2.6:1.2-event-mouse
E: BACKSPACE=guess
E: DEVLINKS=/dev/input/by-path/pci-0000:39:00.0-usb-0:2.6:1.2-event-mouse /dev/input/by-id/usb-Logitech_USB_Receiver-if02-event-mouse
E: DEVNAME=/dev/input/event12
E: DEVPATH=/devices/pci0000:00/0000:00:1c.4/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb3/3-2/3-2.6/3-2.6:1.2/0003:046D:C52B.0016/0003:046D:4063.0017/input/input95/event12
E: ID_BUS=usb
E: ID_INPUT=1
E: ID_INPUT_KEY=1
E: ID_INPUT_KEYBOARD=1
E: ID_INPUT_MOUSE=1
E: ID_MODEL=USB_Receiver
E: ID_MODEL_ENC=USB\x20Receiver
E: ID_MODEL_ID=c52b
E: ID_PATH=pci-0000:39:00.0-usb-0:2.6:1.2
E: ID_PATH_TAG=pci-0000_39_00_0-usb-0_2_6_1_2
E: ID_REVISION=2405
E: ID_SERIAL=Logitech_USB_Receiver
E: ID_TYPE=hid
E: ID_USB_DRIVER=usbhid
E: ID_USB_INTERFACES=:030101:030102:030000:
E: ID_USB_INTERFACE_NUM=02
E: ID_VENDOR=Logitech
E: ID_VENDOR_ENC=Logitech
E: ID_VENDOR_ID=046d
E: LIBINPUT_DEVICE_GROUP=3/46d/4063:usb-0000:39:00.0-2
E: MAJOR=13
E: MINOR=76
E: SUBSYSTEM=input
E: TAGS=:power-switch:
E: USEC_INITIALIZED=53787500514
E: XKBLAYOUT=de
E: XKBMODEL=pc105
```
in which I don't see the property.

**Comment by whot** [comment]
Right, so the issue is the hwdb match not applying. Looking at this again: this is the receiver, not the mouse itself, so you must not use a `name:*` match. [1] If you do that, your rule would apply to all Logitech Unifying devices. On newish-enough kernels (we're talking about at least one/two years here) the devices should be listed with the device's vid/pid, not just the receiver's vid/pid. What kernel version are you on? Is this the only event node you get? 

The MX Anywhere 2 has a special wheel feature for high-res scrolling, the patches for that were only just merged into the kernel and libinput does not yet have support for it. AFAIU the default is low-res scrolling though which is the normal scroll wheel behaviour.

[1] concidentally, this *may* be the source of your issues. I keep forgetting how the hwdb matching works precisely but it's not always as obvious as you'd like it to be, so it might be that once you change this to use the device name, it'll just work.

**Comment by whot** [comment]
`E: LIBINPUT_DEVICE_GROUP=3/46d/4063:usb-0000:39:00.0-2` -> this suggests that the PID that libinput sees is `4063`, not `c52b`. Try that, then we can figure out why `ID_MODEL_ID` is different to `PRODUCT` where libinput gets this bit from (that'd be set on the parent device, `/sys/class/input/eventXX/device`)

**Comment by whot** [comment]
Bought a MX Anywhere 2s yesterday (because it's impossible to get the 2 here) and this match definitely works:
```
mouse:usb:v046dp406a:name:Logitech MX Anywhere 2S:
```
which means that yours should work too. Are you running `udevadm hwdb --update` after the hwdb updates? If this still doesn't work, do this: in the `70-mouse.rules` file, add this line before the other `KERNELS` checks
```
KERNELS=="input*",  ENV{FOO}="hwdb 'mouse:$env{ID_BUS}:v$attr{id/vendor}p$attr{id/product}:name:$attr{name}:'"
```
This is the same bit that usually calls the builtin, but we're setting it to an env variable so we can get the arguments to the builtin.

Run `udevadm test /sys/class/input/eventXYZ` and note the output of the FOO property. Should be something like this:
```
FOO=hwdb 'mouse:usb:v046dp406a:name:Logitech MX Anywhere 2S:'
```
That is going to be your hwdb match. Fix whatever is incorrect in your entry, then run `udevadm hwdb --test 'mouse:usb:v046dp406a:name:Logitech MX Anywhere 2S:'` to verify it works.

**Comment by cbruegg** [comment]
@whot Thanks, that helped getting the rule to show up for my mouse. Unfortunately, scrolling speed is unchanged. Does systemd use some other property for my mouse? See [this](https://gist.github.com/cbruegg/235712e46151200baf4b4240f46cc7ee).

**Comment by BjoernDaase** [comment]
@whot Now, I see the property. For some reason I always need a reboot after each change including executing `udevadm hwdb --update` and `udevadm trigger /dev/input/event*`...
Here is the output with the property:
```
P: /devices/pci0000:00/0000:00:1c.4/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb3/3-2/3-2.6/3-2.6:1.2/0003:046D:C52B.0018/0003:046D:4063.0019/input/input119/event7
N: input/event7
S: input/by-id/usb-Logitech_USB_Receiver-if02-event-mouse
S: input/by-path/pci-0000:39:00.0-usb-0:2.6:1.2-event-mouse
E: BACKSPACE=guess
E: DEVLINKS=/dev/input/by-id/usb-Logitech_USB_Receiver-if02-event-mouse /dev/input/by-path/pci-0000:39:00.0-usb-0:2.6:1.2-event-mouse
E: DEVNAME=/dev/input/event7
E: DEVPATH=/devices/pci0000:00/0000:00:1c.4/0000:03:00.0/0000:04:02.0/0000:39:00.0/usb3/3-2/3-2.6/3-2.6:1.2/0003:046D:C52B.0018/0003:046D:4063.0019/input/input119/event7
E: ID_BUS=usb
E: ID_INPUT=1
E: ID_INPUT_KEY=1
E: ID_INPUT_KEYBOARD=1
E: ID_INPUT_MOUSE=1
E: ID_MODEL=USB_Receiver
E: ID_MODEL_ENC=USB\x20Receiver
E: ID_MODEL_ID=c52b
E: ID_PATH=pci-0000:39:00.0-usb-0:2.6:1.2
E: ID_PATH_TAG=pci-0000_39_00_0-usb-0_2_6_1_2
E: ID_REVISION=2405
E: ID_SERIAL=Logitech_USB_Receiver
E: ID_TYPE=hid
E: ID_USB_DRIVER=usbhid
E: ID_USB_INTERFACES=:030101:030102:030000:
E: ID_USB_INTERFACE_NUM=02
E: ID_VENDOR=Logitech
E: ID_VENDOR_ENC=Logitech
E: ID_VENDOR_ID=046d
E: LIBINPUT_DEVICE_GROUP=3/46d/4063:usb-0000:39:00.0-2
E: MAJOR=13
E: MINOR=71
E: MOUSE_WHEEL_CLICK_ANGLE=30
E: SUBSYSTEM=input
E: TAGS=:power-switch:
E: USEC_INITIALIZED=118259844518
E: XKBLAYOUT=de
E: XKBMODEL=pc105
```

Running `udevadm hwdb --test 'mouse:usb:v046dp406a:name:Logitech MX Anywhere 2:'` verifies that it works and it returns `MOUSE_WHEEL_CLICK_ANGLE=30`.

Anyway, I have the same behavior as @cbruegg, the scroll speed did not change and `libinput debug-events` still returns the following line when scrolling
` event7   POINTER_AXIS      +5.93s	vert 15.00* horiz 0.00 (wheel)`.

**Comment by PeterFeicht** [comment]
This is the `udevadm info` output for a Logitech MX Anywhere 2 connected via Bluetooth, which seems to have a different ID than the USB-connected one:

```
P: /devices/virtual/misc/uhid/0005:046D:B013.0003/input/input31/event20
N: input/event20
E: DEVNAME=/dev/input/event20
E: DEVPATH=/devices/virtual/misc/uhid/0005:046D:B013.0003/input/input31/event20
E: ID_BUS=bluetooth
E: ID_INPUT=1
E: ID_INPUT_MOUSE=1
E: LIBINPUT_DEVICE_GROUP=5/46d/b013:9C:B6:D0:D1:07:3A
E: MAJOR=13
E: MINOR=84
E: MOUSE_WHEEL_CLICK_ANGLE=20
E: MOUSE_WHEEL_CLICK_COUNT=18
E: SUBSYSTEM=input
E: USEC_INITIALIZED=12909968105
```

FWIW, creating a local hwdb file and triggering an update worked immediately on my system.

**Comment by whot** [comment]
@BjoernDaase grab libinput from git master please, it now logs the wheel click angle it uses in `--verbose`, that should help debugging things.

@PeterFeicht you don't need `MOUSE_WHEEL_CLICK_COUNT` if it's the same as the click angle anyway (18 * 20 == 360), that one was just introduced for non-integer click angles.

**Comment by PeterFeicht** [comment]
@whot I interpreted the comments in the hwdb file to mean both should be specified, I guess _clients_ doesn't refer to me adding another entry :-)

**Comment by BjoernDaase** [comment]
@whot Alright, I installed it. How am I supposed to see the output now? Which command has to be executed?

**Comment by whot** [comment]
@PeterFeicht yeah, "clients"  refers to whatever reads the udev hwdb so unless you have some strange hobbies, that's mostly libinput :) specifying both is redundant in your case, so only specifying one is better. and `MOUSE_WHEEL_CLICK_ANGLE` came first, so it's the one that needs to be specified.

@BjoernDaase `libinput debug-events --verbose` should have an output line specifying what libinput sees the udev property as

**Comment by BjoernDaase** [comment]
@whot That's everything I see related to my mouse:

```
event18 - Logitech MX Anywhere 2: is tagged by udev as: Keyboard Mouse
event18 - Logitech MX Anywhere 2: device is a pointer
event18 - Logitech MX Anywhere 2: device is a keyboard
-event18  DEVICE_ADDED     Logitech MX Anywhere 2            seat0 default group8  cap:kp left scroll-nat scroll-button
```

**Comment by whot** [comment]
For some reason libinput doesn't see the udev property then, but I have no idea why if it's listed in the udevadm info output. For reference, the message is logged before the "is tagged by udev" message. 

```
event9  - wheel: vert click angle: 20.00
event9  - Logitech MX Anywhere 2S: is tagged by udev as: Keyboard Mouse
event9  - Logitech MX Anywhere 2S: device is a pointer
event9  - Logitech MX Anywhere 2S: device is a keyboard
```

Sorry, this isn't something I can debug from here with the log, you'll have to investigate what exactly is going on here, it looks like all the pieces are in place.

**Comment by poettering** [comment]
Hmm, so can we close this here? it appears the property is properly attached to the device now in udev, and what#s left to debug is outside of the scope of udev, no?

**Comment by whot** [comment]
@poettering no PR for that device was submitted yet (afaik), the one linked above was for a different device. @BjoernDaase feel free to submit a PR for what you have now, if the property shows up in the udevadm output but doesn't work we can continue to discuss this here or in a libinput issue. But no need to keep the bug open once the PR is merged.

**Comment by BjoernDaase** [comment]
@whot The problem is, that I do not have a working situation here and so I'm not completely sure that's completely correct. I hope I'll have a little bit more time during Christmas and can dig deeper into the problem.

**Comment by poettering** [comment]
So I figure the original issue has been solved, i.e. that the hwdb entry was not applied. let's close the issue then. 

If there's a libinput issue, please file a bug against libinput. And if there's an hwdb entry to add to our repository, please post a PR for it, happy to merge this. But I think this 6y old issue we can close.
